### ISSUE TITLE
InputLanguage.LayoutName can return the wrong value when multiple layouts for a single language exist

### ISSUE THREAD

**Issue by AlexVallat** [dotnet/winforms#4345]
* .NET Core Version: 5
* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**
When there are multiple keyboard layouts of the same language for a single language installed, InputLanguage.LayoutName returns the same value for all of them.

**Expected behavior:**
The correct name of each keyboard layout should be returned

**Minimal repro:**
In Windows 10, add the Bulgarian language (just basic typing), then add both the Bulgarian and Bulgarian (Phonetic Traditional) keyboard layouts. Use InputLanguage.InstalledInputLanguages to iterate over both layouts, note that the InputLanguage for both returns the same value of "Bulgarian"

**Analysis**
Looking at the code, this is because it's looking up substitutions in HKCU\Substitutes but matching on language alone: https://github.com/dotnet/winforms/blob/a07615a219d84f7af24f639dc84f18427083f754/src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs#L180

If you look in the registry key, you can see values for both 00000402 and d0010402 which would both match Bulgarian language (0402), and therefore the same substitution is used for both layouts. I believe the fix is probably to search all substitutions and then check the Layout Id value of all target substitutes to see which one matches the correct layout Id.

**Comment by RussKie** [comment]
Do you have a fix in mind?

**Comment by AlexVallat** [comment]
I am not 100% sure on how the Keyboard Layout\Substitutes registry key is intended to function, so this could benefit from the attention of someone with inside knowledge of this area Windows.

Otherwise, my best guess at a fix just by inspecting the code would be that instead of finding just a single substitution that matches it finds all matching substitutions and puts them in a list. Then remove the loop at https://github.com/dotnet/winforms/blob/a07615a219d84f7af24f639dc84f18427083f754/src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs#L204 and just use the second loop at https://github.com/dotnet/winforms/blob/a07615a219d84f7af24f639dc84f18427083f754/src/System.Windows.Forms/src/System/Windows/Forms/InputLanguage.cs#L235 but searching first for the discovered substitution targets before searching for everything else.

Unit tests for this will be a problem as it depends heavily on the windows environment.

**Comment by weltkante** [comment]
I can confirm that the problem comes from the `Substitutes` key and how WinForms handles it.

I don't think this is actionable without advice from someone who knows how this works. As far as I can tell these registry keys were never documented, and WinForms took a dependency on how the registry keys _used to work_ but apparently they no longer work that way.

Its been a while that LCIDs are no longer unique, and all headers have big warnings to no longer use LCID-based code. The matching algorithm WinForms uses for substitutions (based solely on language ID) probably is outdated.

Randomly guessing a replacement algorithm until it works for your scenario, without knowing the design of these redirections and what guarantees Windows gives going forward, will be a maintenance nightmare. Also note that the implementation must work on all supported platforms, it may very well be that Windows 10 behaves differently from older OS versions, or that the behavior changed in a major OS update.

If no advice/documentation from another team can be obtained the only sensible thing I see is either removing this API, or leaving it in its current broken state. Note that [the documentation says](https://docs.microsoft.com/en-us/windows/win32/inputdev/about-keyboard-input)

> Applications are generally not expected to manipulate input languages directly. Instead, the user sets up language and layout combinations, then switches among them.

The win32 API provides no methods which operate on anything but the current keyboard layout, a strong indicator that Windows understands this as a user setting which applications should not mess with. The fact that WinForms provides an API anyways probably was a mistake, since this API was not maintained and now is broken.

TLDR: my recommendation is
- seek advice from team owning input languages on how to maintain this API
- if none can be obtained, either remove this API or leave it in is historic broken state, since it can clearly not be supported

**Comment by weltkante** [comment]
Having debugged everything a bit I have the suspicion that the `Substitutions` registry section may be wrong to use in the first place, if you already have a input language handle the substitution might already have taken place? In that case separating the handle into language and layout id and matching that against the `Keyboard Layouts` registry section would be sufficient. I still couldn't find any documentation so I wouldn't recommend using this algorithm without official confirmation of how things work.

**Comment by AlexVallat** [comment]
> In that case taking the high part of the layout handle should be entirely sufficient for matching against the Keyboard Layouts registry section, the Substitutions section could be ignored.

That was my first thought too, but it fails for the case when you have a Language with a Keyboard Layout for a different language under it. For example, under English language add a Portuguese keyboard. The handle is then 0x08160809, which is not going to do you any good without the `Substitutes` key. In this case, the existing behaviour would match a `Substitutes` key of `d0010809` by language part (0809), and get the resulting value of `00000816`, which when looked up in keyboard layouts gets you Portuguese (with no layout ID). I've no idea what the `d001` part encodes.

I think, by inspection, if the top nibble of the handle is `0xf0` then the next byte is the keyboard layout id, and can then be looked up without substitution, but otherwise the high word is a language ID. Whether that can just be used directly to look up the keyboard or if the `Substitutes` key is required in that situation is, again, unknown.

So yeah, I agree with you, someone who knows how Keyboard Layouts works internally in Windows really needs to take a look at this to get a proper fix, rather than just a "patch it up so it works for the cases I've thought of, on the version of Windows I've got in front of me" fix.

**Comment by weltkante** [comment]
> The handle is then 0x08160809, which is not going to do you any good without the Substitutes key.

Thats why I'm saying don't look at the `Substitutes` section at all. You no longer have enough information to use that registry section.

The high part of the HKL seems to contain the Keyboard Layout ID (if 0xF000 bits are set) or the Keyboard Language ID (if not set). Both can be looked up directly in the `Keyboard Layouts` section without going through the Substitues. (Note that some layouts in the layout section have no layout id, those are apparently the primary layouts of their language, and will use a language id instead of a layout id.) The low part of the HKL is the culture in which the keyboard layout was defined and you don't have to remap that through the substitutes, as the mapping result already is in the high part. 

I've tested this extensively on my machine yesterday, even for weird stuff like languages which have no language id (Windows generates a placeholder language ID), and it seems to work. The problem is (a) will it work on older supported Windows versions and (b) is it safe to rely on it not changing with the next Windows update.

In general Windows is allowed to (and does) change implementation, so relying on undocumented implementation details _will_ break at some point, its exactly the reason why the WinForms implementation broke now: the original authors derived an algorithm outside the win32 API, which at one point may have been correct, but never updated when Windows changed the implementation.

**Comment by AlexVallat** [comment]
Another nasty little edge case to consider:

Adding a keyboard layout with a layout ID to a language that it doesn't match. For example, for English language, add keyboard layouts for Portuguese and for Bulgarian (Latin)

The handles are:
Portuguese: `0x8160809`
United Kingdom: `0x08090809`
Bulgarian (Latin): `0xf0040809`

I don't think there's any way we can get get the layout name for Bulgarian (Latin) without going through the `Substitutes` key, which has the following entries:

`d0010809` = `00000816`
`d0020809` = `00010402`

So clearly the first one is for Portuguese and the second for Bulgarian, but I can't figure out any correspondence from `0xf0040809` to `d0020809`, and just the suspicious nature of 1 and 2 point to them just being sequential for the `0x0809` language. Which leads me back to my original idea of iterate over all the substitutes for the language and match against layout ID (if the high part of the handle is an ID, marked by 0xf), or just against language (if the high part of the handle is a language)

**Comment by weltkante** [comment]
> I don't think there's any way we can get get the layout name for Bulgarian (Latin) without going through the Substitutes key

You take the high part of the HKL which is `0xF004`, realize its `Layout ID = 0004`, lookup the single `Keyboard Layout` with that `Layout ID`. You are probably missing the fact that `Layout ID`s seem to be unqiue. You probably want to build a cache of them so you don't have to reiterate them all the time, though.

> I can't figure out any correspondence from `0xf0040809` to `d0020809`

You are not supposed to, `d0020809` is the preload key, it doesn't appear in the `Keyboard Layout` section. There is no reverse mapping of a loaded keyboard layout to which user-configured entry caused it to be loaded, simply because a `HKL` can be loaded by a program as well. The Windows API WinForms is calling returns a `HKL` which in no way is related to the preload/substitutes section, except that the preload list determines how the list is initially populated.

> Which leads me back to my original idea of iterate over all the substitutes for the language and match against layout ID (if the high part of the handle is an ID, marked by 0xf), or just against language (if the high part of the handle is a language)

Thats what I've been saying, mostly. You can build a cache for Layout IDs since they seem to be unique. The substitutes-mapping already has been performed, there is no reason WinForms should be doing it again, especially since it doesn't have the key from the preload section (i.e. WinForms doesn't know which keyboard layout entry it is in the list of multiple keyboard layouts for a language). 

But all this discussion is basically speculating about implementation details, and WinForms is not going to be able to maintain them across WinForms releases unless they get the confirmation from the responsible team. This is something you can do in your own library if you need to, but (in my opinion) for integration in WinForms it needs long-term support.

**Comment by AlexVallat** [comment]
> You are probably missing the fact that Layout IDs seem to be unqiue

Ah, yes, I did not consider that. I assumed that Layout ID was only unique per language, and therefore to fully specify a keyboard layout you needed both its layout ID (if any) and language.

I agree that whether this is true or not is just speculation, and WinForms should have either documentation or internal confirmation from someone inside Windows to implement a proper fix.

**Comment by RussKie** [comment]
I got some advice from Windows team, posting it verbatim with re-styling (adjusting for the lack of colouring) below:

<blockquote>

The substitution matching logic in the .NET code does look flawed as it will match on the language ID alone.
The high word of GetKeyboardLayoutList does incorporate the LayoutId info found in the key.

### From my environment:

`GetKeyboardLayout(0)`: 

    0xf0a3 0402  <- Low nibble of the High word contains layout ID information (a3) – low word contains language ID (0402)

`GetKeyboardLayoutName`: 

    0002 0402  <-  high word contains the KLID (0002) – this is for current layout which is Bulgarian Phonetic in my case 
                                                      – per the comments in the .NET code, this API won’t work for your purposes.

`GetKeyboardLayoutList`:

    0x04090409  <- high word contains keyboard layout information, low word contains language information  - in this case, English layout, English language
    0x040c040c
    0x04110411
    0x04120412
    0x08040804
    0xf0020409   <- in this case high word matches the layout entry with layout ID of 02, English language
    0xf0010409
    0xf0aa0402
    0xf0a30402  <- .NET is using this API to return a list of language / layouts.  Low word is Bulgarian, high word low nibble contains layout information.   Layout ID is 0xa3 for Bulgarian language.

### From registry:

System uses preload to load initial set of keyboard layouts for the user

`HKEY_CURRENT_USER\keyboard layout\preload`

    2    REG_SZ    00000402   <- bulgarian is 2nd on my list
    3    REG_SZ    d0010402
    1    REG_SZ    00000409
    4    REG_SZ    d0010409
    5    REG_SZ    d0020409
    6    REG_SZ    0000040c
    7    REG_SZ    00000411
    8    REG_SZ    00000412
    9    REG_SZ    00000804

`HKEY_CURRENT_USER\keyboard layout\Substitutes`

    00000402    REG_SZ    00020402  <- system uses substitute table to map to KLID  (not always, sometimes there isn’t a substitute) – with current .NET logic this would match for both 0xf0aa0402 and 0xf0a30402
    d0010402    REG_SZ    00030402
    d0010409    REG_SZ    00010409
    d0020409    REG_SZ    00020409

`HKEY_LOCAL_MACHINE\system\currentcontrolset\control\keyboard layouts\00020402`  <- stored by KLID in this registry location

| Name | Type | Data |
|---|---|----|
| Layout Display Name  |  REG_EXPAND_SZ  |  `@%SystemRoot%\system32\input.dll,-5173` |
| Layout File |   REG_SZ  |   `KBDBGPH.DLL` |
| Layout Id  |  REG_SZ  |  `00A3` <- Layout ID is here – so this is the entry that has the Layout Display Name 
| Layout Text  |  REG_SZ  |  `Bulgarian (Phonetic)`


In this case I don’t think you should be using the substitute table at all.   I believe it is irrelevant to what you’re trying to do, which is map the results from `GetKeyboardLayoutList` to a list of keyboard layout descriptions.
The keyboard layout name information is found in `HKEY_LOCAL_MACHINE\system\currentcontrolset\control\keyboard layouts`.  I think .NET code should use the Low Word of `GetKeyboardLayoutList`, and iterate through all entries in `HKLM\SystemCurrentControlSet\Control\Keyboard Layouts` looking for matching low word.   When low word is matching, if `high word & 0xF000` then go look for match with `Layout Id`.  If `High word & 0xF000 == zero`, use the high word to look up the entry in `HKEY_LOCAL_MACHINE\system\currentcontrolset\control\keyboard layouts`.  

For example, if I had an entry like `0x040C0409` in `GetKeyboardLayoutList`, this would indicate English language with a French keyboard layout.   I think you would want to look up the entry under `0000040C` in this case.

</blockquote>

**Comment by AlexVallat** [comment]
This reply does suggest that we can't treat layout IDs as unique, as the algorithm proposed is to search first for a matching low word (language) before checking for a matching layout Id. But it is not explicit about the reason for that, and perhaps the answerer just hadn't considered this, as I had not?

**Comment by RussKie** [comment]
A lot of our implementations go back to mid 90s, and behaviours we take for granted now either haven't existed, were in very rudimentary forms or were quite different back then.
As per the advice, our implementation is flawed, and it is likely needs an adjustment to match the Windows behaviour.

**Comment by DJm00n** [comment]
Hi all!

I have written the code in C++ accodring to the [comment](https://github.com/dotnet/winforms/issues/4345#issuecomment-759161693). Seems simple enought and working so far:

```cpp
// Returns KLID string
// Same as GetKeyboardLayoutName but for any HKL
// https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/windows-language-pack-default-values
std::wstring GetKLIDFromHKL(HKL hkl)
{
    std::wstring inputLocaleId;

    WORD langId = LOWORD(hkl);
    WORD deviceId = HIWORD(hkl);

    if ((deviceId & 0xf000) == 0xf000) // Search by 'Layout ID'
    {
        WORD layoutId = deviceId & 0x0fff;
        DWORD index = 0;
        WCHAR subKey[MAX_PATH];
        HKEY key;
        CHECK_EQ(::RegOpenKeyW(HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Keyboard Layouts", &key), ERROR_SUCCESS);
        while (::RegEnumKeyW(key, index, subKey, MAX_PATH) == ERROR_SUCCESS)
        {
            WCHAR layoutIdStr[MAX_PATH];
            DWORD layoutIdStrSize = sizeof(layoutIdStr);
            if (::RegGetValueW(key, subKey, L"Layout Id", RRF_RT_REG_SZ, nullptr, layoutIdStr, &layoutIdStrSize) == ERROR_SUCCESS)
            {
                if (layoutId == std::stoul(layoutIdStr, nullptr, 16))
                {
                    inputLocaleId.assign(subKey);
                    DBGPRINT("Found KLID 0x%ls by layoutId=0x%04x", inputLocaleId.c_str(), layoutId);

                    break;
                }
            }
            ++index;
        }
        CHECK_EQ(::RegCloseKey(key), ERROR_SUCCESS);
    }
    else // Direct request to keyboard layout
    {
        if (deviceId != 0 && langId != deviceId)
            langId = deviceId;

        WCHAR subKey[MAX_PATH];
        std::swprintf(subKey, MAX_PATH, L"%08X", langId);
        inputLocaleId.assign(subKey);

        // Just debug check that layout really exist
        HKEY key, key2;
        CHECK_EQ(::RegOpenKeyW(HKEY_LOCAL_MACHINE, L"SYSTEM\\CurrentControlSet\\Control\\Keyboard Layouts", &key), ERROR_SUCCESS);
        CHECK_EQ(::RegOpenKeyW(key, subKey, &key2), ERROR_SUCCESS);
        CHECK_EQ(::RegCloseKey(key2), ERROR_SUCCESS);
        CHECK_EQ(::RegCloseKey(key), ERROR_SUCCESS);

        DBGPRINT("Found KLID 0x%ls by langId=0x%04x", inputLocaleId.c_str(), langId);
    }

    CHECK(!inputLocaleId.empty());

    return inputLocaleId;
}
```

It could be easely ported to WinForms.

**Comment by RussKie** [comment]
Thank you for sharing @DJm00n.
